app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/OffboardRouterActivityKt.kt

```diff
@@ -0,0 +1,194 @@
+import com.mapbox.navigation.route.offboard.MapboxOffboardRouter
+import com.mapbox.navigation.route.offboard.router.NavigationRoute
+import com.mapbox.navigation.utils.extensions.ifNonNull
+import com.mapbox.services.android.navigation.testapp.R
+import com.mapbox.services.android.navigation.testapp.utils.Utils
+import com.mapbox.services.android.navigation.ui.v5.route.NavigationMapRoute
+import com.mapbox.services.android.navigation.v5.utils.extensions.mapToDirectionsRoute
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+import kotlinx.android.synthetic.main.activity_mock_navigation.*
+import timber.log.Timber
+
+class OffboardRouterActivityKt : AppCompatActivity(),
+    OnMapReadyCallback,
+    MapboxMap.OnMapClickListener,
+    DirectionsSession.RouteObserver {
+
+    private var mapboxMap: MapboxMap? = null
+
+    private lateinit var navigationMapRoute: NavigationMapRoute
+    private var directionsSession: DirectionsSession? = null
+    private var route: DirectionsRoute? = null
+    private var origin: Point? = null
+    private var destination: Point? = null
+    private var waypoint: Point? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_mock_navigation)
+
+        mapView.onCreate(savedInstanceState)
+        mapView.getMapAsync(this)
+
+        newLocationFab.setOnClickListener { onNewLocationClick() }
+    }
+
+    private fun newOrigin() {
+        mapboxMap?.let { map ->
+            clearMap()
+            val latLng = Utils.getRandomLatLng(doubleArrayOf(-77.1825, 38.7825, -76.9790, 39.0157))
+            origin = Point.fromLngLat(latLng.longitude, latLng.latitude)
+            map.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 12.0))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    override fun onMapReady(mapboxMap: MapboxMap) {
+        this.mapboxMap = mapboxMap
+        this.mapboxMap?.addOnMapClickListener(this)
+        mapboxMap.setStyle(Style.MAPBOX_STREETS) {
+            navigationMapRoute = NavigationMapRoute(mapView, mapboxMap)
+            Snackbar.make(findViewById(R.id.container), "Tap map to place waypoint", Snackbar.LENGTH_LONG).show()
+            newOrigin()
+        }
+    }
+
+    override fun onMapClick(point: LatLng): Boolean {
+        when {
+            destination == null -> {
+                destination = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+                findRoute()
+            }
+            waypoint == null -> {
+                waypoint = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+                findRoute()
+            }
+            else -> {
+                Toast.makeText(this, "Only 2 waypoints supported", Toast.LENGTH_LONG).show()
```

Comment:
why we have limit for waypoints? Api doesn't have it

Reply:
This is just an example and was configured to accept only 2 waypoints. Of course, the API accepts multiple waypoints 😉

Reply:
better to leave a comment. When I first met it I supposed that our SDK apply 2 waypoints only. It's confused a little